Collect multiple missing fields on struct pattern into a single diagnostic

When confusing a Fn-like struct with a regular struct in an `if let` pattern, the output leaves a lot to be desired:
```rust
struct S(usize, usize, usize, usize);

fn main() {
    if let S { a, b, c, d } = S(1, 2, 3, 4) {
        println!("hi");
    }
}
```
<details><summary>
<a href="https://play.rust-lang.org/?gist=7480b80ff35037ef78d4727e80e7abca&version=nightly">Current output</a></summary>

```
error[E0026]: struct `S` does not have a field named `a`
 --> src/main.rs:4:16
  |
4 |     if let S { a, b, c, d } = S(1, 2, 3, 4) {
  |                ^ struct `S` does not have field `a`
 
error[E0026]: struct `S` does not have a field named `b`
 --> src/main.rs:4:19
  |
4 |     if let S { a, b, c, d } = S(1, 2, 3, 4) {
  |                   ^ struct `S` does not have field `b`

error[E0026]: struct `S` does not have a field named `c`
 --> src/main.rs:4:22
  |
4 |     if let S { a, b, c, d } = S(1, 2, 3, 4) {
  |                      ^ struct `S` does not have field `c`

error[E0026]: struct `S` does not have a field named `d`
 --> src/main.rs:4:25
  |
4 |     if let S { a, b, c, d } = S(1, 2, 3, 4) {
  |                         ^ struct `S` does not have field `d`

error[E0027]: pattern does not mention field `0`
 --> src/main.rs:4:12
  |
4 |     if let S { a, b, c, d } = S(1, 2, 3, 4) {
  |            ^^^^^^^^^^^^^^^^ missing field `0`
  |
  = note: trying to match a tuple variant with a struct variant pattern

error[E0027]: pattern does not mention field `1`
 --> src/main.rs:4:12
  |
4 |     if let S { a, b, c, d } = S(1, 2, 3, 4) {
  |            ^^^^^^^^^^^^^^^^ missing field `1`
  |
  = note: trying to match a tuple variant with a struct variant pattern

error[E0027]: pattern does not mention field `2`
 --> src/main.rs:4:12
  |
4 |     if let S { a, b, c, d } = S(1, 2, 3, 4) {
  |            ^^^^^^^^^^^^^^^^ missing field `2`
  |
  = note: trying to match a tuple variant with a struct variant pattern

error[E0027]: pattern does not mention field `3`
 --> src/main.rs:4:12
  |
4 |     if let S { a, b, c, d } = S(1, 2, 3, 4) {
  |            ^^^^^^^^^^^^^^^^ missing field `3`
  |
  = note: trying to match a tuple variant with a struct variant pattern

error: aborting due to 8 previous errors
```
</details>

We should probably collect all these errors into a single diagnostic error:

```
error[E0026]: struct `S` does not have a field named `a`, `b`, `c` or `d`
 --> src/main.rs:4:16
  |
4 |     if let S { a, b, c, d } = S(1, 2, 3, 4) {
  |                ^  ^  ^  ^ struct `S` does not have these fields
  = note: trying to match a tuple variant with a struct variant pattern

error[E0027]: pattern does not mention fields `0`, `1`, `2`, or `3`
 --> src/main.rs:4:12
  |
4 |     if let S { a, b, c, d } = S(1, 2, 3, 4) {
  |            ^^^^^^^^^^^^^^^^ missing fields `0`, `1`, `2`, and `3`
  |
  = note: trying to match a tuple variant with a struct variant pattern
  = help: you probably meant to use tuple struct syntax:
  |
4 |     if let S (a, b, c, d) = S(1, 2, 3, 4) {
  |            ^^^^^^^^^^^^^^
```

Even better if there was only one diagnostic.

--- a/src/librustc_typeck/check/_match.rs
+++ b/src/librustc_typeck/check/_match.rs
@@ -938,14 +938,14 @@ https://doc.rust-lang.org/reference/types.html#trait-objects");
         }
 
         if inexistent_fields.len() > 0 {
-            let field_names = if inexistent_fields.len() == 1 {
-                format!("a field named `{}`", inexistent_fields[0].1)
+            let (field_names, t, plural) = if inexistent_fields.len() == 1 {
+                (format!("a field named `{}`", inexistent_fields[0].1), "this", "")
             } else {
-                format!("fields named {}",
-                        inexistent_fields.iter()
+                (format!("fields named {}",
+                         inexistent_fields.iter()
                             .map(|(_, name)| format!("`{}`", name))
                             .collect::<Vec<String>>()
-                            .join(", "))
+                            .join(", ")), "these", "s")
             };
             let spans = inexistent_fields.iter().map(|(span, _)| *span).collect::<Vec<_>>();
             let mut err = struct_span_err!(tcx.sess,
@@ -955,12 +955,13 @@ https://doc.rust-lang.org/reference/types.html#trait-objects");
                                            kind_name,
                                            tcx.item_path_str(variant.did),
                                            field_names);
-            for (span, name) in &inexistent_fields {
+            if let Some((span, _)) = inexistent_fields.last() {
                 err.span_label(*span,
-                               format!("{} `{}` does not have field `{}`",
+                               format!("{} `{}` does not have {} field{}",
                                        kind_name,
                                        tcx.item_path_str(variant.did),
-                                       name));
+                                       t,
+                                       plural));
             }
             if tcx.sess.teach(&err.get_code().unwrap()) {
                 err.note(

--- a/src/librustc_typeck/check/_match.rs
+++ b/src/librustc_typeck/check/_match.rs
@@ -993,13 +993,25 @@ https://doc.rust-lang.org/reference/types.html#trait-objects");
                 tcx.sess.span_err(span, "`..` cannot be used in union patterns");
             }
         } else if !etc {
-            for field in variant.fields
+            let unmentioned_fields = variant.fields
                 .iter()
-                .filter(|field| !used_fields.contains_key(&field.name)) {
+                .map(|field| field.name)
+                .filter(|field| !used_fields.contains_key(&field))
+                .collect::<Vec<_>>();
+            if unmentioned_fields.len() > 0 {
+                let field_names = if unmentioned_fields.len() == 1 {
+                    format!("field `{}`", unmentioned_fields[0])
+                } else {
+                    format!("fields {}",
+                            unmentioned_fields.iter()
+                                .map(|name| format!("`{}`", name))
+                                .collect::<Vec<String>>()
+                                .join(", "))
+                };
                 let mut diag = struct_span_err!(tcx.sess, span, E0027,
-                                                "pattern does not mention field `{}`",
-                                                field.name);
-                diag.span_label(span, format!("missing field `{}`", field.name));
+                                                "pattern does not mention {}",
+                                                field_names);
+                diag.span_label(span, format!("missing {}", field_names));
                 if variant.ctor_kind == CtorKind::Fn {
                     diag.note("trying to match a tuple variant with a struct variant pattern");
                 }

--- a/src/librustc_typeck/check/_match.rs
+++ b/src/librustc_typeck/check/_match.rs
@@ -904,6 +904,7 @@ https://doc.rust-lang.org/reference/types.html#trait-objects");
         // Keep track of which fields have already appeared in the pattern.
         let mut used_fields = FxHashMap();
 
+        let mut inexistent_fields = vec![];
         // Typecheck each field.
         for &Spanned { node: ref field, span } in fields {
             let field_ty = match used_fields.entry(field.name) {
@@ -927,40 +928,53 @@ https://doc.rust-lang.org/reference/types.html#trait-objects");
                             self.field_ty(span, f, substs)
                         })
                         .unwrap_or_else(|| {
-                            let mut err = struct_span_err!(
-                                tcx.sess,
-                                span,
-                                E0026,
-                                "{} `{}` does not have a field named `{}`",
-                                kind_name,
-                                tcx.item_path_str(variant.did),
-                                field.name
-                            );
-                            err.span_label(span,
-                                           format!("{} `{}` does not have field `{}`",
-                                                   kind_name,
-                                                   tcx.item_path_str(variant.did),
-                                                   field.name));
-                            if tcx.sess.teach(&err.get_code().unwrap()) {
-                                err.note(
-                                    "This error indicates that a struct pattern attempted to \
-                                     extract a non-existent field from a struct. Struct fields \
-                                     are identified by the name used before the colon : so struct \
-                                     patterns should resemble the declaration of the struct type \
-                                     being matched.\n\n\
-                                     If you are using shorthand field patterns but want to refer \
-                                     to the struct field by a different name, you should rename \
-                                     it explicitly."
-                                );
-                            }
-                            err.emit();
-
+                            inexistent_fields.push((span, field.name));
                             tcx.types.err
                         })
                 }
             };
 
             self.check_pat_walk(&field.pat, field_ty, def_bm, true);
+        }
+
+        if inexistent_fields.len() > 0 {
+            let field_names = if inexistent_fields.len() == 1 {
+                format!("a field named `{}`", inexistent_fields[0].1)
+            } else {
+                format!("fields named {}",
+                        inexistent_fields.iter()
+                            .map(|(_, name)| format!("`{}`", name))
+                            .collect::<Vec<String>>()
+                            .join(", "))
+            };
+            let spans = inexistent_fields.iter().map(|(span, _)| *span).collect::<Vec<_>>();
+            let mut err = struct_span_err!(tcx.sess,
+                                           spans,
+                                           E0026,
+                                           "{} `{}` does not have {}",
+                                           kind_name,
+                                           tcx.item_path_str(variant.did),
+                                           field_names);
+            for (span, name) in &inexistent_fields {
+                err.span_label(*span,
+                               format!("{} `{}` does not have field `{}`",
+                                       kind_name,
+                                       tcx.item_path_str(variant.did),
+                                       name));
+            }
+            if tcx.sess.teach(&err.get_code().unwrap()) {
+                err.note(
+                    "This error indicates that a struct pattern attempted to \
+                     extract a non-existent field from a struct. Struct fields \
+                     are identified by the name used before the colon : so struct \
+                     patterns should resemble the declaration of the struct type \
+                     being matched.\n\n\
+                     If you are using shorthand field patterns but want to refer \
+                     to the struct field by a different name, you should rename \
+                     it explicitly."
+                );
+            }
+            err.emit();
         }
 
         // Require `..` if struct has non_exhaustive attribute.

--- a/src/test/compile-fail/struct-pat-derived-error.rs
+++ b/src/test/compile-fail/struct-pat-derived-error.rs
@@ -16,10 +16,8 @@ struct a {
 impl a {
     fn foo(&self) {
         let a { x, y } = self.d; //~ ERROR no field `d` on type `&a`
-        //~^ ERROR struct `a` does not have a field named `x`
-        //~^^ ERROR struct `a` does not have a field named `y`
-        //~^^^ ERROR pattern does not mention field `b`
-        //~^^^^ ERROR pattern does not mention field `c`
+        //~^ ERROR struct `a` does not have fields named `x`, `y`
+        //~| ERROR pattern does not mention fields `b`, `c`
     }
 }
 

--- a/src/test/ui/error-codes/E0026-teach.stderr
+++ b/src/test/ui/error-codes/E0026-teach.stderr
@@ -2,7 +2,7 @@ error[E0026]: struct `Thing` does not have a field named `z`
   --> $DIR/E0026-teach.rs:21:23
    |
 LL |         Thing { x, y, z } => {}
-   |                       ^ struct `Thing` does not have field `z`
+   |                       ^ struct `Thing` does not have this field
    |
    = note: This error indicates that a struct pattern attempted to extract a non-existent field from a struct. Struct fields are identified by the name used before the colon : so struct patterns should resemble the declaration of the struct type being matched.
            

--- a/src/test/ui/error-codes/E0026.stderr
+++ b/src/test/ui/error-codes/E0026.stderr
@@ -2,7 +2,7 @@ error[E0026]: struct `Thing` does not have a field named `z`
   --> $DIR/E0026.rs:19:23
    |
 LL |         Thing { x, y, z } => {}
-   |                       ^ struct `Thing` does not have field `z`
+   |                       ^ struct `Thing` does not have this field
 
 error: aborting due to previous error
 

--- a/src/test/ui/missing-fields-in-struct-pattern.rs
+++ b/src/test/ui/missing-fields-in-struct-pattern.rs
@@ -12,6 +12,8 @@ struct S(usize, usize, usize, usize);
 
 fn main() {
     if let S { a, b, c, d } = S(1, 2, 3, 4) {
+    //~^ ERROR struct `S` does not have fields named `a`, `b`, `c`, `d` [E0026]
+    //~^ ERROR pattern does not mention fields `0`, `1`, `2`, `3` [E0027]
         println!("hi");
     }
 }

--- a/src/test/ui/missing-fields-in-struct-pattern.rs
+++ b/src/test/ui/missing-fields-in-struct-pattern.rs
@@ -13,7 +13,7 @@ struct S(usize, usize, usize, usize);
 fn main() {
     if let S { a, b, c, d } = S(1, 2, 3, 4) {
     //~^ ERROR struct `S` does not have fields named `a`, `b`, `c`, `d` [E0026]
-    //~^ ERROR pattern does not mention fields `0`, `1`, `2`, `3` [E0027]
+    //~| ERROR pattern does not mention fields `0`, `1`, `2`, `3` [E0027]
         println!("hi");
     }
 }

--- a/src/test/ui/missing-fields-in-struct-pattern.rs
+++ b/src/test/ui/missing-fields-in-struct-pattern.rs
@@ -1,0 +1,17 @@
+// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+struct S(usize, usize, usize, usize);
+
+fn main() {
+    if let S { a, b, c, d } = S(1, 2, 3, 4) {
+        println!("hi");
+    }
+}

--- a/src/test/ui/missing-fields-in-struct-pattern.stderr
+++ b/src/test/ui/missing-fields-in-struct-pattern.stderr
@@ -1,0 +1,22 @@
+error[E0026]: struct `S` does not have fields named `a`, `b`, `c`, `d`
+  --> $DIR/missing-fields-in-struct-pattern.rs:14:16
+   |
+LL |     if let S { a, b, c, d } = S(1, 2, 3, 4) {
+   |                ^  ^  ^  ^ struct `S` does not have field `d`
+   |                |  |  |
+   |                |  |  struct `S` does not have field `c`
+   |                |  struct `S` does not have field `b`
+   |                struct `S` does not have field `a`
+
+error[E0027]: pattern does not mention fields `0`, `1`, `2`, `3`
+  --> $DIR/missing-fields-in-struct-pattern.rs:14:12
+   |
+LL |     if let S { a, b, c, d } = S(1, 2, 3, 4) {
+   |            ^^^^^^^^^^^^^^^^ missing fields `0`, `1`, `2`, `3`
+   |
+   = note: trying to match a tuple variant with a struct variant pattern
+
+error: aborting due to 2 previous errors
+
+Some errors occurred: E0026, E0027.
+For more information about an error, try `rustc --explain E0026`.

--- a/src/test/ui/missing-fields-in-struct-pattern.stderr
+++ b/src/test/ui/missing-fields-in-struct-pattern.stderr
@@ -2,11 +2,7 @@ error[E0026]: struct `S` does not have fields named `a`, `b`, `c`, `d`
   --> $DIR/missing-fields-in-struct-pattern.rs:14:16
    |
 LL |     if let S { a, b, c, d } = S(1, 2, 3, 4) {
-   |                ^  ^  ^  ^ struct `S` does not have field `d`
-   |                |  |  |
-   |                |  |  struct `S` does not have field `c`
-   |                |  struct `S` does not have field `b`
-   |                struct `S` does not have field `a`
+   |                ^  ^  ^  ^ struct `S` does not have these fields
 
 error[E0027]: pattern does not mention fields `0`, `1`, `2`, `3`
   --> $DIR/missing-fields-in-struct-pattern.rs:14:12

--- a/src/test/ui/numeric-fields.stderr
+++ b/src/test/ui/numeric-fields.stderr
@@ -10,7 +10,7 @@ error[E0026]: struct `S` does not have a field named `0x1`
   --> $DIR/numeric-fields.rs:17:17
    |
 LL |         S{0: a, 0x1: b, ..} => {}
-   |                 ^^^^^^ struct `S` does not have field `0x1`
+   |                 ^^^^^^ struct `S` does not have this field
 
 error: aborting due to 2 previous errors
 

--- a/src/test/ui/type-check/issue-41314.stderr
+++ b/src/test/ui/type-check/issue-41314.stderr
@@ -2,7 +2,7 @@ error[E0026]: variant `X::Y` does not have a field named `number`
   --> $DIR/issue-41314.rs:17:16
    |
 LL |         X::Y { number } => {} //~ ERROR does not have a field named `number`
-   |                ^^^^^^ variant `X::Y` does not have field `number`
+   |                ^^^^^^ variant `X::Y` does not have this field
 
 error[E0027]: pattern does not mention field `0`
   --> $DIR/issue-41314.rs:17:9

--- a/src/test/ui/union/union-fields-2.stderr
+++ b/src/test/ui/union/union-fields-2.stderr
@@ -52,7 +52,7 @@ error[E0026]: union `U` does not have a field named `c`
   --> $DIR/union-fields-2.rs:28:19
    |
 LL |     let U { a, b, c } = u; //~ ERROR union patterns should have exactly one field
-   |                   ^ union `U` does not have field `c`
+   |                   ^ union `U` does not have this field
 
 error: union patterns should have exactly one field
   --> $DIR/union-fields-2.rs:28:9

